syntax = "proto3";

package envoy.config.filter.network.zookeeper_proxy.v1alpha1;

import "google/protobuf/wrappers.proto";

import "udpa/annotations/migrate.proto";
import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.envoyproxy.envoy.config.filter.network.zookeeper_proxy.v1alpha1";
option java_outer_classname = "ZookeeperProxyProto";
option java_multiple_files = true;
option go_package = "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/zookeeper_proxy/v1alpha1";
option (udpa.annotations.file_migrate).move_to_package =
    "envoy.extensions.filters.network.zookeeper_proxy.v3";
option (udpa.annotations.file_status).package_version_status = FROZEN;

// [#protodoc-title: ZooKeeper proxy]
// ZooKeeper Proxy :ref:`configuration overview <config_network_filters_zookeeper_proxy>`.
// [#extension: envoy.filters.network.zookeeper_proxy]

message ZooKeeperProxy {
  // The human readable prefix to use when emitting :ref:`statistics
  // <config_network_filters_zookeeper_proxy_stats>`.
  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];

  // [#not-implemented-hide:] The optional path to use for writing ZooKeeper access logs.
  // If the access log field is empty, access logs will not be written.
  string access_log = 2;

  // Messages — requests, responses and events — that are bigger than this value will
  // be ignored. If it is not set, the default value is 1Mb.
  //
  // The value here should match the jute.maxbuffer property in your cluster configuration:
  //
  // https://zookeeper.apache.org/doc/r3.4.10/zookeeperAdmin.html#Unsafe+Options
  //
  // if that is set. If it isn't, ZooKeeper's default is also 1Mb.
  google.protobuf.UInt32Value max_packet_bytes = 3;

  // List of latency thresholds for opcodes to decide the fast/slow requests and calculate
  // the SLI used for error budget calculation.
  // If the threshold of one opcode is not set, it will fallback to use the default
  // threshold.
  // If the default threshold is not set, it will be set as 100 millisecond.
  repeated LatencyThreshold latency_thresholds = 4;
}

message LatencyThreshold {
  oneof opcode {
    google.protobuf.UInt32Value default = 1 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value connect = 2 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value create = 3 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value delete = 4 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value exists = 5 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value getData = 6 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value setData = 7 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value getAcl = 8 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value setAcl = 9 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value get_children = 10 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value sync = 11 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value ping = 12 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value get_children2 = 13 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value check = 14 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value multi = 15 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value create2 = 16 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value reconfig = 17 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value check_watches = 18 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value remove_watches = 19 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value create_container = 20 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value create_ttl = 21 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value close = 22 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value set_auth = 23 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value set_watches = 24 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value get_ephemerals = 25 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value get_all_children_number = 26 [(validate.rules).uint32.gt = 0];
    google.protobuf.UInt32Value set_watches2 = 27 [(validate.rules).uint32.gt = 0];
  }
}
